File: src/assets/nodes_builtin/http.js
// HTTP 请求节点插件（type 45）：方法/URL/请求体/超时。
// params：method/url/timeout 走通用 widget；body 按 bodyType 联动出 JSON/表单/纯文本编辑器（custom）。
EdFW.register({
  type: "45", label: "HTTP",     icon: "🌐", category: "io",
  section: "HTTP 请求",
  params: [
    { key: "method", label: "方法", widget: "select", options: ["GET", "POST", "PUT", "DELETE"],
      get(n) { return n.data.inputs?.apiInfo?.method || "GET"; },
      set(n, v) { let a = n.data.inputs.apiInfo || (n.data.inputs.apiInfo = {}); a.method = v; if (v === "GET") { const b = n.data.inputs.body || (n.data.inputs.body = {}); b.bodyType = "EMPTY"; } } },
    { key: "url", label: "URL", widget: "input", tip: "https://... 支持 {{字段名}} 占位符",
      get(n) { return n.data.inputs?.apiInfo?.url || ""; },
      set(n, v) { let a = n.data.inputs.apiInfo || (n.data.inputs.apiInfo = {}); a.url = v; } },
    { key: "body", label: "请求体", widget: "custom",
      get(n) { const bd = n.data.inputs?.body || {}; return bd.bodyType || "EMPTY"; }, set() {},
      html(n) {
        const api = n.data.inputs?.apiInfo || {};
        if (api.method !== "POST" && api.method !== "PUT") return "";
        const bd = n.data.inputs?.body || {};
        let h = `<label>请求体类型 <select onchange="NODEP_HTTP_set('bodyType',this.value)">` +
                ["EMPTY|无", "JSON|JSON", "FORM_URLENCODED|表单(URL编码)", "RAW_TEXT|纯文本"]
                  .map(s => { const [v, t] = s.split("|"); return `<option value="${v}"${(bd.bodyType || "EMPTY") === v ? " selected" : ""}>${t}</option>`; }).join("") +
                `</select></label>`;
        if (bd.bodyType === "JSON") {
          h += `<label>JSON 体 <textarea onchange="NODEP_HTTP_set('bodyJson',this.value)" style="min-height:80px">${ext((bd.bodyData || {}).json)}</textarea></label>` +
               `<div style="font-size:10px;color:#8a9099">支持 {{输入字段名}} 占位符引用上游数据</div>`;
        }
        if (bd.bodyType === "RAW_TEXT") {
          h += `<label>文本体 <textarea onchange="NODEP_HTTP_set('bodyText',this.value)" style="min-height:60px">${ext((bd.bodyData || {}).rawText)}</textarea></label>`;
        }
        if (bd.bodyType === "FORM_URLENCODED") {
          h += `<label>表单字段</label>`;
          const fields = (bd.bodyData || {}).formURLEncoded || [];
          fields.forEach((p, pi) => {
            h += `<div style="display:flex;gap:3px;margin:2px 0">` +
                 `<input value="${ext(p.name)}" placeholder="键" onchange="NODEP_HTTP_form(${pi},'name',this.value)" style="flex:1;font-size:10px">` +
                 `<input value="${ext(p.input?.value?.content)}" placeholder="值" onchange="NODEP_HTTP_form(${pi},'val',this.value)" style="flex:2;font-size:10px">` +
                 `<button class="del" onclick="NODEP_HTTP_formDel(${pi})">×</button></div>`;
          });
          h += `<button onclick="NODEP_HTTP_formAdd()" style="font-size:10px">+ 字段</button>`;
        }
        return h;
      } },
    { key: "timeout", label: "超时(秒)", widget: "number",
      get(n) { return (n.data.inputs?.setting || {}).timeout ?? 15; },
      set(n, v) { let s = n.data.inputs.setting || (n.data.inputs.setting = {}); s.timeout = parseInt(v) || 15; } },
  ],
});
// HTTP 编辑器全局
function NODEP_HTTP_set(field, v) {
  const n = findN(selNode); const d = n.data.inputs;
  if (field === "method") { let a = d.apiInfo || (d.apiInfo = {}); a.method = v; if (v === "GET") { const b = d.body || (d.body = {}); b.bodyType = "EMPTY"; } }
  else if (field === "url") { let a = d.apiInfo || (d.apiInfo = {}); a.url = v; }
  else if (field === "bodyType") { let b = d.body || (d.body = {}); b.bodyType = v; if (v === "JSON") { b.bodyData = b.bodyData || {}; b.bodyData.json = b.bodyData.json || ""; } if (v === "RAW_TEXT") { b.bodyData = b.bodyData || {}; b.bodyData.rawText = b.bodyData.rawText || ""; } if (v === "FORM_URLENCODED") { b.bodyData = b.bodyData || {}; b.bodyData.formURLEncoded = b.bodyData.formURLEncoded || []; } }
  else if (field === "bodyJson") { let b = d.body || (d.body = {}); b.bodyData = b.bodyData || {}; b.bodyData.json = v; }
  else if (field === "bodyText") { let b = d.body || (d.body = {}); b.bodyData = b.bodyData || {}; b.bodyData.rawText = v; }
  else if (field === "timeout") { let s = d.setting || (d.setting = {}); s.timeout = parseInt(v) || 15; }
  showProps(); renderAll();
}
function NODEP_HTTP_form(pi, field, v) {
  const n = findN(selNode); const bd = (n.data.inputs.body || (n.data.inputs.body = {}));
  bd.bodyData = bd.bodyData || {}; bd.bodyData.formURLEncoded = bd.bodyData.formURLEncoded || [];
  const p = bd.bodyData.formURLEncoded[pi]; if (!p) return;
  if (field === "name") p.name = v;
  else { p.input = p.input || { type: "string", value: { type: "literal", content: "" } }; p.input.value.content = v; }
}
function NODEP_HTTP_formAdd() {
  const n = findN(selNode); const bd = (n.data.inputs.body || (n.data.inputs.body = {}));
  bd.bodyData = bd.bodyData || {}; bd.bodyData.formURLEncoded = bd.bodyData.formURLEncoded || [];
  bd.bodyData.formURLEncoded.push({ name: "", input: { type: "string", value: { type: "literal", content: "" } } });
  showProps(); renderAll();
}
function NODEP_HTTP_formDel(pi) {
  const n = findN(selNode); const bd = n.data.inputs.body || {};
  if (bd.bodyData?.formURLEncoded) bd.bodyData.formURLEncoded.splice(pi, 1);
  showProps(); renderAll();
}

File: src/assets/nodes_builtin/http.py
"""HTTP 请求节点插件（type 45）：method/url/headers/params/body/auth。
URL/JSON 体支持 {{变量名}} 模板（输入参数引用）。
"""

PARAMS = [
    {"key": "method",   "type": "string", "required": True, "enum": ["GET", "POST", "PUT", "DELETE"],
     "desc": "HTTP 方法"},
    {"key": "url",      "type": "string", "required": True,
     "desc": "目标 URL；{{输入字段名}} 占位符"},
    {"key": "body",     "type": "object", "required": False,
     "desc": "POST/PUT 请求体：{bodyType: EMPTY|JSON|FORM_URLENCODED|RAW_TEXT, bodyData:{...}}"},
    {"key": "timeout",  "type": "number", "required": False, "default": 15,
     "desc": "超时秒数"},
]

import json

from workflow_node_api import resolve_input_params, resolve_value, render_template


def _handle_http(node: dict, ctx) -> dict:
    import urllib.parse
    import urllib.request
    import urllib.error

    inputs = node.get("data", {}).get("inputs", {})
    params = resolve_input_params(inputs.get("inputParameters", []), ctx)   # URL/body 的 {{变量名}} 引用这些输入
    api = inputs.get("apiInfo", {}) or {}
    method = (api.get("method") or "GET").upper()
    url = render_template(api.get("url", ""), params)

    kv = {}
    for p in inputs.get("params", []) or []:
        kv[p.get("name")] = resolve_value(p.get("input"), ctx)
    if kv:
        url += ("&" if "?" in url else "?") + urllib.parse.urlencode(kv, doseq=True)

    headers = {}
    for p in inputs.get("headers", []) or []:
        headers[p.get("name")] = str(resolve_value(p.get("input"), ctx))

    auth = inputs.get("auth") or {}
    if auth.get("authOpen") and auth.get("authType") == "bearer":
        for p in ((auth.get("authData") or {}).get("bearerTokenData") or []):
            if p.get("name") == "token":
                headers["Authorization"] = "Bearer " + str(resolve_value(p.get("input"), ctx))

    body = inputs.get("body") or {}
    bt = body.get("bodyType")
    bd = body.get("bodyData") or {}
    data = None
    if bt == "JSON":
        data = render_template(bd.get("json", ""), params).encode("utf-8")
        headers.setdefault("Content-Type", "application/json")
    elif bt == "RAW_TEXT":
        data = render_template(bd.get("rawText", ""), params).encode("utf-8")
        headers.setdefault("Content-Type", "text/plain")
    elif bt in ("FORM_DATA", "FORM_URLENCODED"):
        fields = bd.get("formURLEncoded") or (bd.get("formData", {}) or {}).get("data") or []
        form = {p.get("name"): str(resolve_value(p.get("input"), ctx)) for p in fields}
        data = urllib.parse.urlencode(form).encode("utf-8")
        headers.setdefault("Content-Type", "application/x-www-form-urlencoded")

    setting = inputs.get("setting") or {}
    timeout = int(setting.get("timeout") or 15)
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            raw_bytes = resp.read()
            code = resp.getcode()
            ctype = resp.headers.get("Content-Type", "")
            hdrs = json.dumps(dict(resp.headers.items()), ensure_ascii=False)
    except urllib.error.HTTPError as e:
        raw_bytes = e.read() if hasattr(e, "read") else str(e).encode()
        code = e.code
        ctype = ""
        hdrs = "{}"
    # 解码：优先 Content-Type 的 charset；utf-8 失败则回退 gbk（中文站点常见）
    charset = "utf-8"
    for _tok in ctype.split(";"):
        if _tok.strip().lower().startswith("charset="):
            charset = _tok.split("=", 1)[1].strip().strip('"')
    try:
        raw = raw_bytes.decode(charset) if isinstance(raw_bytes, (bytes, bytearray)) else str(raw_bytes)
    except (LookupError, UnicodeDecodeError):
        try:
            raw = raw_bytes.decode("gbk", errors="replace")
        except Exception:
            raw = str(raw_bytes)
    except Exception as e:
        return {"outputs": {"body": f"[HTTP 失败] {type(e).__name__}: {e}", "statusCode": 0, "headers": "{}"},
                "port": None}
    return {"outputs": {"body": raw, "statusCode": code, "headers": hdrs}, "port": None}


def agt_node():
    return {"type": "45", "label": "HTTP", "handler": _handle_http, "catalog": _CATALOG}

# ===== 节点目录条目（list_workflow_nodes / query_workflow_node 动态聚合自插件声明）=====
_CATALOG = {"name": "HTTP 请求 (HTTP)", "desc": "发起 HTTP 请求（GET/POST/PUT/DELETE），支持 headers/params/body/auth 配置，URL 和 body 中可用 {{}} 模板引用上游输出", "xml": "<!-- HTTP 请求节点 -->\n<node id=\"250001\" type=\"http\">\n  <!-- API 信息：method 和 url（url 中可用 {{变量}} 模板） -->\n  <param name=\"method\" literal=\"POST\">POST</param>\n  <param name=\"url\"><![CDATA[https://api.example.com/v1/chat/completions]]></param>\n\n  <!-- 请求头 -->\n  <param name=\"Content-Type\" literal=\"application/json\" header=\"true\">application/json</param>\n  <param name=\"Authorization\" header=\"true\"><![CDATA[Bearer {{api_key}}]]></param>\n\n  <!-- URL 查询参数 -->\n  <param name=\"version\" literal=\"v1\" query=\"true\">v1</param>\n\n  <!-- 模板入参 -->\n  <in name=\"api_key\" ref=\"190001.api_key\"/>\n  <in name=\"body_data\" ref=\"230001.output\"/>\n\n  <!-- 请求体（JSON body） -->\n  <param name=\"bodyType\" literal=\"json\">json</param>\n  <body><![CDATA[{{body_data}}]]></body>\n\n  <!-- 超时和重试 -->\n  <param name=\"timeout\" type=\"integer\">30</param>\n  <param name=\"retryTimes\" type=\"integer\">2</param>\n\n  <out name=\"body\" type=\"string\"/>\n  <out name=\"statusCode\" type=\"integer\"/>\n  <out name=\"headers\" type=\"object\"/>\n</node>\n<!--\n  header=\"true\" 的 param 作为请求头；query=\"true\" 的 param 作为 URL 查询参数\n  body 元素内的 CDATA 为请求体，支持 {{变量}} 模板\n  输出：body（响应体字符串）, statusCode（HTTP 状态码）, headers（响应头JSON对象）\n-->"}
